### PR TITLE
[Bug] Add Operator Kronecker product 

### LINKED ISSUE #168: [Bug] Density_mat and torch.kron
When I try to execute this lines:
```python
rho_0 = density_mat(product_state('0'))
promote_operator(rho_0,0,2)
```
The output is:
```python
RuntimeError: view size is not compatible with input tensor's size and stride (at least one dimension spans across two contiguous subspaces). Use .reshape(...) instead.
```
But if I execute these lines, which are completely equivalent, I don’t get any errors:
```python
rho_0 = DensityMatrix(pyq.Projector(1,'0','0').unitary())
promote_operator(rho_0,0,2)
```
I know this error comes from the function `torch.kron()` because `torch.kron(rho_0.contiguous(),I(0).unitary())` leads to the same error. 

Also, I already had the same error message when i tried to execute:
```python
torch.kron(I(0).dagger(),I(0).unitary())
```
According to this [tread](https://github.com/pytorch/pytorch/issues/54135), add `.contiguous()` resolved the problem, but this time the `permutation` does by `density_mat` is not a `transpose` and even with `.contiguous()` the problem is still there...

### PR DESCRIPTION
**Pull request by EthanObadia**
Resolved the bug report #168 by adding `operator_kron()` function in `utils.py`.
Closed #168.

### CODE REVIEW

**Comment by EthanObadia** [comment]
> @EthanObadia what was the bug?

When I try to execute those lines:
```python
rho_0, rho_1 = density_mat(product_state('0')), density_mat(product_state('1'))
torch.kron(rho_0, rho_1)
```
Or those:
```python
rho_0 = density_mat(product_state('0'))
promote_operator(rho_0,0,2)
```

The output is:
```python
RuntimeError: view size is not compatible with input tensor's size and stride (at least one dimension spans across two contiguous subspaces). Use .reshape(...) instead.
```
So this basically impossible to compute $\rho_{01} = \rho_0 \otimes \rho_1$ or $\rho^{\prime} = X^{prom}\otimes \rho_1$. 

The pyq's operators shape is not convenient for `torch.kron()`, so I code a special one. 
I tried to explain more [here](https://arc.net/l/quote/jfhijttk)

**Comment by EthanObadia** [comment]
After more tests, this version of `operator_kron` needs to change, because it also does the `torch.kron` on the `batch_size`:
```python3
operator_kron(density_mat(product_state('0',2)),X(0).unitary().repeat(1,1,2))
```
Returns a `torch.Size([4, 4, 4])` tensor but we expect to keep the same number of batches so to obtain a `torch.Size([4, 4, 2])` tensor.

File: tests/test_digital.py [hunk 1]
```diff
@@ -342,3 +349,41 @@ def test_operator_product(gate: Primitive) -> None:
     assert torch.allclose(
         op_mul, torch.eye(2**n_qubits, dtype=torch.cdouble).unsqueeze(2).repeat(1, 1, max_batch)
     )
+
+
+@pytest.mark.parametrize("operator,matrix", [(I, IMAT), (X, XMAT), (Z, ZMAT), (Y, YMAT), (H, HMAT)])
+def test_operator_kron(operator: Tensor, matrix: Tensor) -> None:
```
Comment (Roland-djee) [inline tests/test_digital.py:355]: Could you maybe break it into two tests ? It's getting hard to understand what's going on really.

**Review by Roland-djee — COMMENTED** [review]
Thanks @EthanObadia. LGTM. Will approve after addressing comment.

File: tests/test_digital.py (hunk 1 above)
Comment (EthanObadia) [reply tests/test_digital.py:355]: Do you prefer this version ?

```python3
@pytest.mark.parametrize("operator,matrix", [(I, IMAT), (X, XMAT), (Z, ZMAT), (Y, YMAT), (H, HMAT)])
def test_operator_kron(operator: Tensor, matrix: Tensor) -> None:
    n_qubits = torch.randint(low=1, high=5, size=(1,)).item()
    batch_size = torch.randint(low=1, high=5, size=(1,)).item()
    states, krons = [], []
    for batch in range(batch_size):
        state = random_state(n_qubits)
        states.append(state)
        kron = torch.kron(density_mat(state).squeeze(), matrix).unsqueeze(2)
        krons.append(kron)
    input_state = torch.cat(states, dim=n_qubits)
    kron_out = operator_kron(density_mat(input_state), operator(0).dagger())
    assert kron_out.size() == torch.Size([2 ** (n_qubits + 1), 2 ** (n_qubits + 1), batch_size])
    kron_expect = torch.cat(krons, dim=2)
    assert torch.allclose(kron_out, kron_expect)
    assert torch.allclose(
        torch.kron(operator(0).dagger().contiguous(), I(0).unitary()),
        operator_kron(operator(0).dagger(), I(0).unitary()),
    )    


def test_kron_batch() -> None: 
    n_qubits = torch.randint(low=1, high=5, size=(1,)).item()
    batch_size_1 = torch.randint(low=1, high=5, size=(1,)).item()
    batch_size_2 = torch.randint(low=1, high=5, size=(1,)).item()
    max_batch = max(batch_size_1, batch_size_2)
    dm_1 = density_mat(random_state(n_qubits, batch_size_1)) 
    dm_2 = density_mat(random_state(n_qubits, batch_size_2))
    dm_out = operator_kron(dm_1, dm_2)
    assert dm_out.size() == torch.Size([2 ** (2 * n_qubits), 2 ** (2 * n_qubits), max_batch])
    if batch_size_1 > batch_size_2:
        dm_2 = dm_2.repeat(1, 1, batch_size_1)[:, :, :batch_size_1]
    elif batch_size_2 > batch_size_1:
        dm_1 = dm_1.repeat(1, 1, batch_size_2)[:, :, :batch_size_2]
    density_matrices = []
    for batch in range(max_batch):
        density_matrice = torch.kron(dm_1[:, :, batch], dm_2[:, :, batch]).unsqueeze(2)
        density_matrices.append(density_matrice)
    dm_expect = torch.cat(density_matrices, dim=2)
    assert torch.allclose(dm_out, dm_expect)
```

**Pull request merged**